Fixed in luarocks!1550 (merged) but not included in a release yet (latest is v3.9.2).
Patch can be removed once a new luarocks version is released.

See https://github.com/luarocks/luarocks/pull/1550

commit f4b1deab270a94778e1bd8903ff48387b0ed9fbc
Author: leso-kn <info@lesosoftware.com>
Date:   Tue Dec 5 22:05:55 2023 +0100

    fix: Add value specified via --tree to cfg.rocks_trees (#1549)
    
    * Fixes a crash in `fulfill_dependency()` on musl

diff --git a/src/luarocks/cmd.lua b/src/luarocks/cmd.lua
index f1446df..ca79c31 100644
--- a/src/luarocks/cmd.lua
+++ b/src/luarocks/cmd.lua
@@ -83,6 +83,9 @@ do
          if not named then
             local root_dir = fs.absolute_name(args.tree)
             replace_tree(args, root_dir)
+            if (args.deps_mode or cfg.deps_mode) ~= "order" then
+               table.insert(cfg.rocks_trees, 1, { name = "arg", root = root_dir } )
+            end
          end
       elseif args["local"] then
          if fs.is_superuser() then
